File: data-jdbc/src/test/groovy/io/micronaut/data/jdbc/oraclexe/etag/ETagBook.java
package io.micronaut.data.jdbc.oraclexe.etag;

import io.micronaut.data.annotation.*;
import io.micronaut.data.annotation.sql.ETagValue;
import io.micronaut.data.annotation.sql.GeneratedETag;
import io.micronaut.data.annotation.sql.ETaggable;

@ETaggable
@MappedEntity("etag_book")
public record ETagBook(
    @Id
    @GeneratedValue
    Long id,
    String title,

    @Relation(Relation.Kind.EMBEDDED)
    BookDetails bookDetails,

    @GeneratedETag
    String etag) {

    @Embeddable
    public record BookDetails(
        int pages,
        @ETagValue(exclude = true)
        int chapters
    ) {
    }
}

File: doc-examples/jdbc-example-records-python/src/main/python/example/ETagBook.py
# tag::generated-etag[]
from dataclasses import dataclass
from typing import Annotated

from micronaut.data.annotation import Embeddable, GeneratedValue, Id, MappedEntity, Relation
from micronaut.data.annotation.sql import ETagValue, ETaggable, GeneratedETag


@Embeddable
@dataclass(frozen=True)
class BookDetails:
    pages: int
    chapters: Annotated[int, ETagValue(exclude=True)]


@MappedEntity("etag_book")
@ETaggable
@dataclass(frozen=True)
class ETagBook:
    id: Annotated[int | None, Id, GeneratedValue]
    title: str
    bookDetails: Annotated[BookDetails, Relation("EMBEDDED")]
    etag: Annotated[str | None, GeneratedETag]
# end::generated-etag[]
